### PR TITLE
feat(imagefamily): implement GKE release channel awareness

### LINKED ISSUE #330: feat: support GKE Upgrade Release Channels in node image selection
**What would you like to be added**:

Support for GKE Upgrade Release Channels in node image selection. Currently the `@latest` alias always resolves to the **newest OS image** in the GCP public catalog for the cluster's current K8s version, regardless of the cluster's configured release channel (RAPID / REGULAR / STABLE / EXTENDED).

The desired addition is:
- A new `@channel` alias version that follows the cluster's configured release channel — resolving to the channel's `defaultVersion` (from `getServerConfig`) and selecting the newest catalog image for that K8s version.
- Optional explicit per-channel alias versions: `@stable`, `@regular`, `@rapid`, `@extended` — each resolving to the corresponding channel's `defaultVersion` from server config, independent of the cluster's own channel setting.

**Why is this needed**:

GKE release channels control the K8s version upgrade cadence. Each channel is typically on a different K8s minor or patch version at any given time (e.g. STABLE on 1.34.x, RAPID on 1.35.x). Since [PR #280](https://github.com/cloudpilot-ai/karpenter-provider-gcp/pull/280) removed the dependency on NodePool templates for image resolution, Karpenter resolves images directly from the public GCP image catalog using `Discovery().ServerVersion()` as the K8s version filter.

This means:
- `@latest` always uses the cluster's *current running* K8s version, which is correct for most cases.
- However, it picks the **newest image in the catalog** for that version — potentially a newer OS build than what GKE has deployed to cluster nodes via the release channel.
- There is no mechanism for a user to explicitly request images matching a specific release channel's K8s version.

Within the same K8s patch version, OS image sub-versions (GKE internal build numbers, e.g. `gke1396004`) are **not exposed per channel** through any public API — confirmed via exhaustive research (see linked comment). The meaningful differentiation between channels is at the K8s version level.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change (`spec.imageSelectorTerms[].alias` — new valid version tokens)
- [ ] Docs update

### PR DESCRIPTION
**Pull request by dm3ch**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Implements [proposal/0002-release-channel-awareness](proposals/0002-release-channel-awareness.md): adds structured `family`/`channel`/`version` fields to `ImageSelectorTerm`, allowing image selection by GKE release channel or explicit version rather than only by the (now-deprecated) `alias` string.

**New `ImageSelectorTerm` fields:**

- `family` — image OS family (`ContainerOptimizedOS`, `Ubuntu2404`, `Ubuntu2204`)
- `channel` — GKE release channel (`rapid`, `regular`, `stable`, `extended`, or `cluster` to follow the cluster's enrolled channel)
- `version` — explicit version string or `latest`

**How it works:**

- When `channel:` is set, the provider calls `GetServerConfig` (Container API, 30-min cache) to resolve the highest K8s minor version shipped for that channel, then derives the corresponding GCE image via an exact GKE build filter (`gke-{k8sKey}-gke{build}-*`).
- `channel: cluster` reads the cluster's own enrolled channel from `GetCluster`.
- When `version:` is set, the milestone (`125.19216.104.126`) or date (`v20260416`) is used directly; invalid formats are rejected at admission.
- A build miss (no image for the channel's GKE version) surfaces as `ImagesReady=False` with reason `ImageResolutionFailed` rather than silently falling back.

**Ubuntu 2204 support:** `isUsableUbuntu2204Image` handles the 2204 naming convention (no `-amd64-` suffix on amd64 images); arm64 derivation inserts `-arm64-` before the `-v{date}` version suffix.

**`alias` backward compatibility:** The existing `alias` field continues to work unchanged. It is now soft-deprecated — new configurations should use `family`/`channel`/`version`.

**Docs:** New `docs/image-selection.md` (selection mode reference); `docs/image-management.md` rewritten to lead with the new paradigm (alias demoted to a Legacy section).

**GC race fix (included in this PR):** Increased the orphan garbage-collection grace period from 30 s to 3 min. GCE VMs appear in `instances.list` within seconds of the create call, but the lifecycle controller only writes `providerID` back to the NodeClaim after the zone operation completes (30–90 s). The previous 30 s grace was narrow enough to delete a legitimately-launching VM before its NodeClaim was updated, leaving the NodeClaim permanently stuck with `Launched=True` but no backing instance.

**E2E fix (included in this PR):** Dropped `c4a` from all four arm64 provisioning specs. `c4a` machines require `hyperdisk-balanced` but e2e NodeClasses use `pd-balanced` by default, so every `c4a` attempt fails with a GCP 400 before falling through to `t2a`. Removing `c4a` makes the arm64 specs reliable without narrowing what Karpenter itself supports.

#### Related proposal (if applicable):

`proposals/0002-release-channel-awareness.md`

#### Which issue(s) this PR fixes:

Fixes #330 

#### Docs and examples

- [x] `docs/` and `examples/` updated (or this PR does not affect user-facing behaviour, configuration, or APIs)
- [x] `MIGRATION.md` updated (or this PR does not require any migration steps)

#### Special notes for your reviewer:

- `alias` CEL validation rules from the previous release are preserved unchanged for backward compatibility.
- `GetServerConfig` results are cached for 30 minutes to avoid hammering the GKE Container API on every reconcile.
- E2E suite `test/suites/channel-image-selection/` validates channel-based, version-pinned, and `version: latest` selection end-to-end against a live GKE cluster.
- `//nolint:staticcheck` on `term.Alias` accesses in `image.go`: the `Alias` field carries a `// Deprecated:` doc comment so that IDEs and `go vet` surface the deprecation to users. That same annotation causes `staticcheck` (SA1019) to flag our own backward-compat handler; the suppression is intentional — we own both the deprecated field and the code that must keep reading it.
- **GC grace period (30 s → 3 min):** The previous 30 s grace was too short — a VM taking ~35 s to provision would pass the grace window before the lifecycle controller wrote the `providerID` back to the NodeClaim, causing the GC to delete it as an orphan. This was observed in e2e: the `COS / amd64 / spot` provisioning spec failed because the GC deleted the VM at `16:44:15` (35 s after creation) while the lifecycle controller completed at `16:44:15.842` (117 ms later). 3 min covers the full provisioning cycle at the cost of one extra GC cycle before a genuine orphan is collected.
- **`c4a` dropped from arm64 e2e specs:** `c4a` machines are incompatible with `pd-balanced` disk (they require `hyperdisk-balanced`). The provisioning tests now use only `t2a-standard-{2,4}` for arm64 — which is what actually works in `asia-southeast1` with standard NodeClass config. This is a test-only change; Karpenter itself continues to surface `c4a` offerings.
- This PR was prepared with AI assistance (Claude Code). All changes have been reviewed and verified by the author.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR implements GKE release channel awareness for image selection, adding structured `family`/`channel`/`version` fields to `ImageSelectorTerm` as an alternative to the now-deprecated `alias` string. It also includes a GC grace-period fix (30 s → 3 min) to prevent premature orphan deletion of legitimately-launching VMs, and drops `c4a` from arm64 e2e specs to fix flaky test failures.

- **Image selection redesign**: `resolveFamilyTerm` routes through a new `resolveVersion` helper that calls `GetServerConfig` (30-min cache) when `channel:` is set and `GetCluster` for `channel: cluster`, then delegates to per-family `ResolveImages` implementations. A new `resolveExactBuildCOSImage` path uses an exact GKE-build filter and surfaces misses as `imageResolutionError`.
- **Ubuntu 2204 support**: `isUsableUbuntu2204Image` identifies amd64 images by excluding `-arm64-`, and `deriveArm64Image` inserts `-arm64-` before the `-v{date}` suffix for that series.
- **GC grace period increase**: `gcGracePeriod` raised to 3 minutes to cover the full VM provisioning cycle before the lifecycle controller writes `providerID` back to the NodeClaim.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the new channel-based image resolution path is well-structured, imageResolutionError is correctly propagated in all new code paths, and the GC grace-period change is conservative and well-justified.

The core logic — channel resolution, version lookup, image filtering, and error classification — all look correct. Unit tests cover the key resolution scenarios, edge cases (UNSPECIFIED cluster channel, missing builds, semver ordering), and the Ubuntu 2204 naming conventions. The two findings are design-level observations rather than present defects.

pkg/providers/imagefamily/ubuntu.go — the ubuntuVersionRe replacement scope; pkg/apis/v1alpha1/gcenodeclass.go — mixed-family ImageSelectorTerms behaviour in ImageFamily().
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| pkg/providers/imagefamily/image.go | Core dispatch layer; adds resolveFamilyTerm and resolveVersion with correct imageResolutionError wrapping for channel/cluster lookup failures and empty-candidate GCP results. |
| pkg/providers/imagefamily/containeroptimizedos.go | Adds ParseGKEVersion and resolveExactBuildCOSImage for channel-based resolution; build-miss is correctly classified as imageResolutionError; arm64 derivation regex requires exactly 4 COS version groups. |
| pkg/providers/imagefamily/ubuntu.go | Adds Ubuntu 2204 support via release-parameterized struct; isUsableUbuntu2204Image correctly identifies amd64 images by excluding -arm64-; deriveArm64Image handles the 2204 naming convention correctly. |
| pkg/providers/gke/version.go | New file implementing ResolveVersionForChannel; correctly uses semver comparison (not lexicographic) and handles defaultVersion/validVersions fallback; well-tested. |
| pkg/providers/gke/gke.go | Adds GetClusterConfig and GetServerConfig with separate 30-min caches; both paths are straightforward GKE API wrappers. |
| pkg/apis/v1alpha1/gcenodeclass.go | Adds Family/Channel/Version fields to ImageSelectorTerm with comprehensive CEL validation; ImageFamily() extended to handle new family terms mapping Ubuntu2404/2204 to ImageFamilyUbuntu for downstream compatibility. |
| pkg/controllers/nodeclaim/garbagecollection/controller.go | GC grace period increased from 30 s to 3 min; well-documented rationale; isOrphaned logic is unchanged and correct. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant NC as NodeClass Reconciler
    participant IP as ImageProvider
    participant GKE as GKE Provider
    participant GCP as GCP Compute API

    NC->>IP: List(nodeClass)
    IP->>IP: hash(ImageSelectorTerms) → cache key
    alt cache hit
        IP-->>NC: cached Images
    else cache miss
        loop each ImageSelectorTerm
            alt term.Alias set
                IP->>IP: resolveAliasTerm(term)
            else term.ID set
                IP->>IP: resolveIDTerm(term)
            else term.Family set
                IP->>IP: resolveFamilyTerm(term)
                alt term.Version set
                    IP->>IP: "version = term.Version"
                else "term.Channel == cluster"
                    IP->>GKE: GetClusterConfig() [30m cache]
                    GKE-->>IP: cluster.ReleaseChannel.Channel
                    IP->>GKE: GetServerConfig() [30m cache]
                    GKE-->>IP: ServerConfig.Channels
                    IP->>IP: ResolveVersionForChannel
                    IP-->>IP: gkeVersion
                else term.Channel set
                    IP->>GKE: GetServerConfig() [30m cache]
                    GKE-->>IP: ServerConfig.Channels
                    IP->>IP: ResolveVersionForChannel
                    IP-->>IP: gkeVersion
                end
                IP->>IP: provider.ResolveImages(ctx, version)
                alt GKE version format
                    IP->>GCP: "Images.List(gke-k8sKey-gkebuild-*)"
                    GCP-->>IP: matching COS images
                    IP->>IP: pick best, build amd64/arm64/gpu URLs
                else latest or milestone pin
                    IP->>GCP: "Images.List(gke-k8sKey-*)"
                    GCP-->>IP: candidates
                    IP->>IP: pick latest, derive variants
                end
                IP->>GCP: Images.Get(each candidate)
                GCP-->>IP: exists or 404
                IP->>IP: imageResolutionError if all absent
            end
        end
        IP->>IP: cache.Set(key, result, 5m)
        IP-->>NC: Images
    end
```
</details>

<a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Apkg%2Fproviders%2Fimagefamily%2Fubuntu.go%3A37%0AThe%20%60ubuntuVersionRe%60%20replacement%20in%20%60ResolveImages%60%20uses%20%60ReplaceAllStringFunc%60%2C%20which%20replaces%20**every**%20non-overlapping%20match%20of%20%60-v%5Cd%2B%28-%7C%24%29%60.%20For%20current%20GKE%20Ubuntu%20image%20names%20there%20is%20only%20one%20match%2C%20but%20an%20image%20name%20segment%20like%20%60-v1-35-%60%20%28if%20naming%20ever%20used%20a%20%60v%60-prefixed%20minor%20version%29%20would%20also%20be%20replaced%2C%20silently%20producing%20a%20wrong%20image%20name.%20The%20%60deriveArm64Image%60%20function%20has%20the%20same%20exposure.%20Anchoring%20the%20replacement%20to%20the%20last%20occurrence%2C%20or%20at%20minimum%20stopping%20after%20the%20first%20match%2C%20would%20make%20the%20intent%20explicit.%0A%0A%60%60%60suggestion%0A%2F%2F%20ubuntuVersionRe%20matches%20the%20version%20token%20at%20the%20end%20of%20a%20GKE%20Ubuntu%20image%20name%0A%2F%2F%20%28e.g.%20%22-v20260416%22%20or%20%22-v20260416-%22%29.%20The%20pattern%20is%20intentionally%20anchored%20to%20the%0A%2F%2F%20last%20occurrence%20so%20that%20numeric%20segments%20earlier%20in%20the%20name%20%28e.g.%20%222404%22%2C%20%221%22%2C%20%2235%22%29%0A%2F%2F%20are%20never%20replaced.%0Avar%20ubuntuVersionRe%20%3D%20regexp.MustCompile%28%60-v%28%5Cd%7B8%7D%29%28-%7C%24%29%60%29%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Apkg%2Fapis%2Fv1alpha1%2Fgcenodeclass.go%3A386-394%0A**Mixed-family%20%60ImageSelectorTerms%60%20produces%20ambiguous%20%60ImageFamily%28%29%60**%0A%0A%60ImageFamily%28%29%60%20now%20returns%20on%20the%20first%20%60family%3A%60%20term%20it%20finds.%20If%20a%20NodeClass%20lists%20both%20COS%20and%20Ubuntu%20terms%20%E2%80%94%20which%20CEL%20validation%20does%20not%20prevent%20%E2%80%94%20the%20first%20term's%20OS%20family%20determines%20kubelet%20env%20patching%2C%20node%20taints%2C%20and%20GPU%20logic%20for%20ALL%20nodes%20provisioned%20from%20that%20class%2C%20regardless%20of%20which%20image%20is%20actually%20selected.%20For%20example%2C%20a%20term%20list%20of%20%60%5B%7Bfamily%3A%20ContainerOptimizedOS%2C%20version%3A%20latest%7D%2C%20%7Bfamily%3A%20Ubuntu2404%2C%20version%3A%20latest%7D%5D%60%20would%20always%20configure%20nodes%20as%20COS%20even%20when%20an%20Ubuntu%20image%20is%20chosen.%0A%0AConsider%20adding%20a%20term-list-level%20CEL%20rule%20enforcing%20that%20all%20%60family%3A%60%20terms%20in%20a%20single%20%60ImageSelectorTerms%60%20use%20the%20same%20OS%20family%20group%2C%20or%20document%20this%20limitation%20clearly.%0A%0A&pr=342&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=3">Fix All in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Apkg%2Fproviders%2Fimagefamily%2Fubuntu.go%3A37%0AThe%20%60ubuntuVersionRe%60%20replacement%20in%20%60ResolveImages%60%20uses%20%60ReplaceAllStringFunc%60%2C%20which%20replaces%20**every**%20non-overlapping%20match%20of%20%60-v%5Cd%2B%28-%7C%24%29%60.%20For%20current%20GKE%20Ubuntu%20image%20names%20there%20is%20only%20one%20match%2C%20but%20an%20image%20name%20segment%20like%20%60-v1-35-%60%20%28if%20naming%20ever%20used%20a%20%60v%60-prefixed%20minor%20version%29%20would%20also%20be%20replaced%2C%20silently%20producing%20a%20wrong%20image%20name.%20The%20%60deriveArm64Image%60%20function%20has%20the%20same%20exposure.%20Anchoring%20the%20replacement%20to%20the%20last%20occurrence%2C%20or%20at%20minimum%20stopping%20after%20the%20first%20match%2C%20would%20make%20the%20intent%20explicit.%0A%0A%60%60%60suggestion%0A%2F%2F%20ubuntuVersionRe%20matches%20the%20version%20token%20at%20the%20end%20of%20a%20GKE%20Ubuntu%20image%20name%0A%2F%2F%20%28e.g.%20%22-v20260416%22%20or%20%22-v20260416-%22%29.%20The%20pattern%20is%20intentionally%20anchored%20to%20the%0A%2F%2F%20last%20occurrence%20so%20that%20numeric%20segments%20earlier%20in%20the%20name%20%28e.g.%20%222404%22%2C%20%221%22%2C%20%2235%22%29%0A%2F%2F%20are%20never%20replaced.%0Avar%20ubuntuVersionRe%20%3D%20regexp.MustCompile%28%60-v%28%5Cd%7B8%7D%29%28-%7C%24%29%60%29%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Apkg%2Fapis%2Fv1alpha1%2Fgcenodeclass.go%3A386-394%0A**Mixed-family%20%60ImageSelectorTerms%60%20produces%20ambiguous%20%60ImageFamily%28%29%60**%0A%0A%60ImageFamily%28%29%60%20now%20returns%20on%20the%20first%20%60family%3A%60%20term%20it%20finds.%20If%20a%20NodeClass%20lists%20both%20COS%20and%20Ubuntu%20terms%20%E2%80%94%20which%20CEL%20validation%20does%20not%20prevent%20%E2%80%94%20the%20first%20term's%20OS%20family%20determines%20kubelet%20env%20patching%2C%20node%20taints%2C%20and%20GPU%20logic%20for%20ALL%20nodes%20provisioned%20from%20that%20class%2C%20regardless%20of%20which%20image%20is%20actually%20selected.%20For%20example%2C%20a%20term%20list%20of%20%60%5B%7Bfamily%3A%20ContainerOptimizedOS%2C%20version%3A%20latest%7D%2C%20%7Bfamily%3A%20Ubuntu2404%2C%20version%3A%20latest%7D%5D%60%20would%20always%20configure%20nodes%20as%20COS%20even%20when%20an%20Ubuntu%20image%20is%20chosen.%0A%0AConsider%20adding%20a%20term-list-level%20CEL%20rule%20enforcing%20that%20all%20%60family%3A%60%20terms%20in%20a%20single%20%60ImageSelectorTerms%60%20use%20the%20same%20OS%20family%20group%2C%20or%20document%20this%20limitation%20clearly.%0A%0A&repo=cloudpilot-ai%2Fkarpenter-provider-gcp&pr=342&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a> <a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22cloudpilot-ai%2Fkarpenter-provider-gcp%22%20on%20the%20existing%20branch%20%22feat%2Frelease-channel-impl%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22feat%2Frelease-channel-impl%22.%0A%0AFix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Apkg%2Fproviders%2Fimagefamily%2Fubuntu.go%3A37%0AThe%20%60ubuntuVersionRe%60%20replacement%20in%20%60ResolveImages%60%20uses%20%60ReplaceAllStringFunc%60%2C%20which%20replaces%20**every**%20non-overlapping%20match%20of%20%60-v%5Cd%2B%28-%7C%24%29%60.%20For%20current%20GKE%20Ubuntu%20image%20names%20there%20is%20only%20one%20match%2C%20but%20an%20image%20name%20segment%20like%20%60-v1-35-%60%20%28if%20naming%20ever%20used%20a%20%60v%60-prefixed%20minor%20version%29%20would%20also%20be%20replaced%2C%20silently%20producing%20a%20wrong%20image%20name.%20The%20%60deriveArm64Image%60%20function%20has%20the%20same%20exposure.%20Anchoring%20the%20replacement%20to%20the%20last%20occurrence%2C%20or%20at%20minimum%20stopping%20after%20the%20first%20match%2C%20would%20make%20the%20intent%20explicit.%0A%0A%60%60%60suggestion%0A%2F%2F%20ubuntuVersionRe%20matches%20the%20version%20token%20at%20the%20end%20of%20a%20GKE%20Ubuntu%20image%20name%0A%2F%2F%20%28e.g.%20%22-v20260416%22%20or%20%22-v20260416-%22%29.%20The%20pattern%20is%20intentionally%20anchored%20to%20the%0A%2F%2F%20last%20occurrence%20so%20that%20numeric%20segments%20earlier%20in%20the%20name%20%28e.g.%20%222404%22%2C%20%221%22%2C%20%2235%22%29%0A%2F%2F%20are%20never%20replaced.%0Avar%20ubuntuVersionRe%20%3D%20regexp.MustCompile%28%60-v%28%5Cd%7B8%7D%29%28-%7C%24%29%60%29%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Apkg%2Fapis%2Fv1alpha1%2Fgcenodeclass.go%3A386-394%0A**Mixed-family%20%60ImageSelectorTerms%60%20produces%20ambiguous%20%60ImageFamily%28%29%60**%0A%0A%60ImageFamily%28%29%60%20now%20returns%20on%20the%20first%20%60family%3A%60%20term%20it%20finds.%20If%20a%20NodeClass%20lists%20both%20COS%20and%20Ubuntu%20terms%20%E2%80%94%20which%20CEL%20validation%20does%20not%20prevent%20%E2%80%94%20the%20first%20term's%20OS%20family%20determines%20kubelet%20env%20patching%2C%20node%20taints%2C%20and%20GPU%20logic%20for%20ALL%20nodes%20provisioned%20from%20that%20class%2C%20regardless%20of%20which%20image%20is%20actually%20selected.%20For%20example%2C%20a%20term%20list%20of%20%60%5B%7Bfamily%3A%20ContainerOptimizedOS%2C%20version%3A%20latest%7D%2C%20%7Bfamily%3A%20Ubuntu2404%2C%20version%3A%20latest%7D%5D%60%20would%20always%20configure%20nodes%20as%20COS%20even%20when%20an%20Ubuntu%20image%20is%20chosen.%0A%0AConsider%20adding%20a%20term-list-level%20CEL%20rule%20enforcing%20that%20all%20%60family%3A%60%20terms%20in%20a%20single%20%60ImageSelectorTerms%60%20use%20the%20same%20OS%20family%20group%2C%20or%20document%20this%20limitation%20clearly.%0A%0A"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a>

<sub>Reviews (7): Last reviewed commit: ["test(provisioning): drop c4a from arm64 ..."](https://github.com/cloudpilot-ai/karpenter-provider-gcp/commit/02a226116181cccc6d09700c5af36f8bfe03bd77) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=32819921)</sub>

### CODE REVIEW

File: charts/karpenter/templates/deployment.yaml [hunk 1]
```diff
@@ -107,6 +107,10 @@ spec:
             - name: VM_MEMORY_OVERHEAD_PERCENT
               value: "{{ . }}"
             {{- end }}
+            {{- with .Values.controller.settings.defaultNodePoolTemplateName }}
```
Comment (dm3ch) [inline charts/karpenter/templates/deployment.yaml:110]: #263 artifact, should desapear after merge/rebase

File: docs/reference/gcenodeclass.md [hunk 2]
```diff
@@ -104,7 +104,7 @@ _Appears in:_
 | --- | --- | --- | --- |
 | `serviceAccount` _string_ | ServiceAccount is the GCP IAM service account email to assign to the instance |  | Pattern: `^[^@]+@(developer\.gserviceaccount\.com\|[^@]+\.iam\.gserviceaccount\.com)$` <br />Optional: \{\} <br /> |
 | `disks` _[Disk](#disk) array_ | Disk defines the disks to attach to the provisioned instance. |  | MaxItems: 10 <br />Optional: \{\} <br /> |
-| `imageSelectorTerms` _[ImageSelectorTerm](#imageselectorterm) array_ | ImageSelectorTerms is a list of or image selector terms. The terms are ORed.<br />Remove or adjust mutual exclusivity rules since there's only one field |  | MaxItems: 30 <br />MinItems: 1 <br />Required: \{\} <br /> |
+| `imageSelectorTerms` _[ImageSelectorTerm](#imageselectorterm) array_ | ImageSelectorTerms is a list of or image selector terms. The terms are ORed. |  | MaxItems: 30 <br />MinItems: 1 <br />Required: \{\} <br /> |
```
Comment (dm3ch) [inline docs/reference/gcenodeclass.md:107]: It seems in our case terms are not ORed but ANDed isn't it family and channel must both match?

**Comment by dm3ch** [comment]
### E2E Test Results

**Cluster:** \`asia-southeast1-b\` | **Commit:** \`23e1edda\` | **Total duration:** ~12m (provisioning), ~5m (drift/gpu/channel)

| Suite | Spec | Result | Duration | Notes |
|---|---|---|---|---|
| Provisioning | COS / amd64 / on-demand | ✅ pass | ~3m | |
| Provisioning | COS / amd64 / spot | ✅ pass | 2m 18s | Previously failing due to GC race — fixed by grace period 30s→3m |
| Provisioning | COS / arm64 / on-demand | ✅ pass | ~3m | |
| Provisioning | COS / arm64 / spot | ✅ pass | ~3m | |
| Provisioning | Ubuntu / amd64 / on-demand | ✅ pass | ~3m | |
| Provisioning | Ubuntu / amd64 / spot | ✅ pass | ~3m | |
| Provisioning | Ubuntu / arm64 / on-demand | ✅ pass | 3m 31s | Previously failing due to `c4a` + `pd-balanced` incompatibility — fixed by dropping `c4a` from arm64 specs |
| Provisioning | Ubuntu / arm64 / spot | ✅ pass | ~3m | |
| Provisioning | Pinned COS version | ✅ pass | ~3m | |
| Provisioning | Pinned Ubuntu version | ✅ pass | ~3m | |
| Provisioning | NodeClass drift | ✅ pass | ~3m | |
| Drift | should replace a node when GCENodeClass metadata changes | ✅ pass | 5m 17s | |
| GPU | should provision a GPU node with correct taint, labels, and allocatable GPU resource | ✅ pass | 3m 8s | |
| ChannelImageSelection | COS / channel: stable / amd64 / on-demand | ✅ pass | ~3m | Channel version verified against server config |
| ChannelImageSelection | Ubuntu2404 / version: latest / amd64 / on-demand | ✅ pass | ~3m | Ubuntu version verified in resolved image |

**Overall: ✅ all pass**

**Review by thameezb — APPROVED** [review]
Great work
